hive-hooks/src/main/java/com/airbnb/reair/hive/hooks/MetastoreAuditLogListener.java

```diff
@@ -0,0 +1,391 @@
+package com.airbnb.reair.hive.hooks;
+
+import com.airbnb.reair.db.DbCredentials;
+import com.airbnb.reair.utils.RetryableTask;
+import com.airbnb.reair.utils.RetryingTaskRunner;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.MetaStoreEventListener;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.events.AddPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.AlterPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.AlterTableEvent;
+import org.apache.hadoop.hive.metastore.events.CreateDatabaseEvent;
+import org.apache.hadoop.hive.metastore.events.CreateTableEvent;
+import org.apache.hadoop.hive.metastore.events.DropDatabaseEvent;
+import org.apache.hadoop.hive.metastore.events.DropPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.DropTableEvent;
+import org.apache.hadoop.hive.ql.hooks.ReadEntity;
+import org.apache.hadoop.hive.ql.hooks.WriteEntity;
+import org.apache.hadoop.hive.ql.hooks.WriteEntity.WriteType;
+import org.apache.hadoop.hive.ql.metadata.Partition;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.log4j.Logger;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Audit logging for the metastore Thrift server. Comprises of a series of
+ * event listeners adding auditing to the various Thrift events.
+ */
+public class MetastoreAuditLogListener extends MetaStoreEventListener {
+  public static Logger LOG = Logger.getLogger(MetastoreAuditLogListener.class);
+
+  // Number of attempts to make.
+  private static final int NUM_ATTEMPTS = 10;
+
+  // Will wait BASE_SLEEP * 2 ^ (attempt no.) between attempts.
+  private static final int BASE_SLEEP = 1;
+
+  public static String DB_USERNAME =
+      "airbnb.reair.metastoreaudit_log.db.username";
+
+  public static String DB_PASSWORD =
+      "airbnb.reair.metastore.audit_log.db.password";
+
+  // Keys for values in hive-site.xml.
+  public static String JDBC_URL_KEY = "airbnb.reair.audit_log.jdbc_url";
```

Comment:
Also, since the JDBC url is separate, we probably also want separate keys for defining the table names.